RecoLocalMuon/GEMRecHit/src/ME0SegAlgoMM.cc

```diff
@@ -0,0 +1,556 @@
+  // erasing the merged ones, rather than the 
+  for(size_t NNN = 0; NNN < seeds.size(); ++NNN) {
+    if(running_meanX[NNN] == 999999.) continue; //skip seeds that have been marked as used up in merging
+    rechits_clusters.push_back(seeds[NNN]);
+  }
+
+  return rechits_clusters; 
+}
+
+
+ME0SegAlgoMM::ProtoSegments 
+ME0SegAlgoMM::chainHits(const EnsambleHitContainer & rechits) {
+
+  ProtoSegments rechits_chains; 
+  EnsambleHitContainer temp;
+  ProtoSegments seeds;
+
+  std::vector <bool> usedCluster;
+
+  // split rechits into subvectors and return vector of vectors:
+  // Loop over rechits
+  // Create one seed per hit
+  for(unsigned int i = 0; i < rechits.size(); ++i) {
+    temp.clear();
+    temp.push_back(rechits[i]);
+    seeds.push_back(temp);
+    usedCluster.push_back(false);
+  }
+
+  // merge chains that are too close ("touch" each other)
+  for(size_t NNN = 0; NNN < seeds.size(); ++NNN) {
+    for(size_t MMM = NNN+1; MMM < seeds.size(); ++MMM) {
+      if(usedCluster[MMM] || usedCluster[NNN]){
+        continue;
+      }
+      // all is in the way we define "good";
+      // try not to "cluster" the hits but to "chain" them;
+      // it does the clustering but also does a better job
+      // for inclined tracks (not clustering them together;
+      // crossed tracks would be still clustered together) 
+      // 22.12.09: In fact it is not much more different 
+      // than the "clustering", we just introduce another
+      // variable in the game - Z. And it makes sense 
+      // to re-introduce Y (or actually wire group mumber)
+      // in a similar way as for the strip number - see
+      // the code below.
+      bool goodToMerge  = isGoodToMerge(seeds[NNN], seeds[MMM]);
+      if(goodToMerge){
+        // merge chains!
+        // merge by adding seed NNN to seed MMM and erasing seed NNN
+
+        // add seed NNN to MMM (lower to larger number)
+        seeds[MMM].insert(seeds[MMM].end(),seeds[NNN].begin(),seeds[NNN].end());
+
+        // mark seed NNN as used
+        usedCluster[NNN] = true;
+        // we have merged a seed (NNN) to the highter seed (MMM) - need to contimue to
+        // next seed (NNN+1)
+        break;
+      }
+
+    }
+  }
+
+  // hand over the final seeds to the output
+  // would be more elegant if we could do the above step with
+  // erasing the merged ones, rather than the
+
+  for(size_t NNN = 0; NNN < seeds.size(); ++NNN) {
+    if(usedCluster[NNN]) continue; //skip seeds that have been marked as used up in merging
+    rechits_chains.push_back(seeds[NNN]);
+  }
+
+  //***************************************************************
+
+      return rechits_chains;
+}
+
+bool ME0SegAlgoMM::isGoodToMerge(EnsambleHitContainer & newChain, EnsambleHitContainer & oldChain) {
+   for(size_t iRH_new = 0;iRH_new<newChain.size();++iRH_new){
+    int layer_new = newChain[iRH_new]->me0Id().layer();     
+    float phi_new = theEnsamble.first->toGlobal(newChain[iRH_new]->localPosition()).phi();
```

Comment:
Either initialize theEnsamble to some valid default value in the constructor or check that theEnsamble.first is not null in each of the five methods where it is dereferenced. We should make sure there is no possibility that theEnsamble.first could be null and cause a segmentation fault when dereferenced.